File: victorious/victorious/UI/Profile/Follow Relationships/VUserIsFollowingDataSource.m
//
//  VUserIsFollowingDataSource.m
//  victorious
//
//  Created by Patrick Lynch on 6/17/15.
//  Copyright (c) 2015 Victorious. All rights reserved.
//

#import "VUserIsFollowingDataSource.h"
#import "VUser.h"
#import "VConstants.h"
#import "MBProgressHUD.h"
#import "victorious-Swift.h"

@interface VUserIsFollowingDataSource ()

@property (nonatomic, strong) NSMutableOrderedSet *followedUsers;

@end

@implementation VUserIsFollowingDataSource

- (instancetype)initWithUser:(VUser *)user
{
    NSParameterAssert( user != nil );
    
    self = [super init];
    if ( self != nil )
    {
        _user = user;
        _pageLoader = [[PageLoader alloc] init];
    }
    return self;
}

- (instancetype)init
{
    NSAssert(NO, @"Use the designated initializer");
    return nil;
}

#pragma mark - VUsersDataSource

- (NSString *)noContentTitle
{
    return self.user.isCurrentUser ? NSLocalizedString( @"NotFollowingTitle", @"" ) : NSLocalizedString( @"ProfileNotFollowingTitle", @"" );
}

- (NSString *)noContentMessage
{
    return self.user.isCurrentUser ? NSLocalizedString( @"NotFollowingMessage", @"" ) : NSLocalizedString( @"ProfileNotFollowingMessage", @"" );
}

- (UIImage *)noContentImage
{
    return [UIImage imageNamed:@"noFollowersIcon"];
}

- (void)refreshWithPageType:(VPageType)pageType completion:(void(^)(BOOL success, NSError *error))completion
{
    [self loadPage:pageType completion:^(NSArray<VUser *> *_Nonnull users, NSError *_Nullable error)
    {
        if ( error == nil )
        {
            if ( pageType == VPageTypeFirst )
            {
                // Start fresh on the first page
                [self.followedUsers removeAllObjects];
            }
            [self.followedUsers addObjectsFromArray:users];
        }
    }];
}

- (NSArray *)users
{
    return self.followedUsers.array;
}

@end

File: victorious/victorious/UI/Profile/Follow Relationships/VUserIsFollowingDataSource.swift
//
//  VUserIsFollowingDataSource.swift
//  victorious
//
//  Created by Patrick Lynch on 12/20/15.
//  Copyright © 2015 Victorious. All rights reserved.
//

import Foundation

@objc final class VUserIsFollowingDataSource: PaginatedDataSource, VUsersDataSource {
    
    let user: VUser
    
    init( user: VUser ) {
        self.user = user
    } 
    
    func users() -> NSOrderedSet {
        return self.visibleItems
    }
    
    func loadUsersWithPageType( pageType: VPageType, completion: (NSError? -> ())? = nil ) {
        let userID = self.user.remoteId.integerValue
        
        self.loadPage( pageType,
            createOperation: {
                return UsersFollowedByUserOperation(userID: userID)
            },
            completion:{ (results, error, cancelled) in
                completion?( error )
            }
        )
    }
    
    // MARK: - VUsersDataSource
    
    func noContentTitle() -> String {
        return self.user.isCurrentUser ? NSLocalizedString( "NotFollowingTitle", comment: "" ) : NSLocalizedString( "ProfileNotFollowingTitle", comment: "" )
    }
    
    func noContentMessage() -> String {
        return self.user.isCurrentUser ? NSLocalizedString( "NotFollowingMessage", comment: "" ) : NSLocalizedString( "ProfileNotFollowingMessage", comment: "" )
    }
    
    func noContentImage() -> UIImage {
        return UIImage(named: "noFollowersIcon" )!
    }
}
